lib/stages/ReadGain.cpp

```diff
@@ -0,0 +1,240 @@
+#include "ReadGain.hpp"
+
+#include "buffer.h"
+#include "bufferContainer.hpp"
+#include "chimeMetadata.h"
+#include "configUpdater.hpp"
+#include "errors.h"
+
+#include <functional>
+#include <utils/visUtil.hpp>
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::configUpdater;
+using kotekan::Stage;
+
+using kotekan::connectionInstance;
+using kotekan::HTTP_RESPONSE;
+using kotekan::restServer;
+
+REGISTER_KOTEKAN_STAGE(ReadGain);
+
+// clang-format off
+
+// Request gain file re-parse with e.g.
+// FRB
+// curl localhost:12048/frb_gain -X POST -H 'Content-Type: appication/json' -d '{"frb_gain_dir":"the_new_path"}'
+// PSR
+// curl localhost:12048/updatable_config/pulsar_gain -X POST -H 'Content-Type: application/json' -d
+// '{"pulsar_gain_dir":["path0","path1","path2","path3","path4","path5","path6","path7","path8","path9"]}'
+//
+// clang-format on
+
+ReadGain::ReadGain(Config& config, const std::string& unique_name,
+                   bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&ReadGain::main_thread, this)) {
+    // Apply config.
+    _num_elements = config.get<uint32_t>(unique_name, "num_elements");
+    _num_beams = config.get<int32_t>(unique_name, "num_beams");
+    scaling = config.get_default<float>(unique_name, "frb_scaling", 1.0);
+    vector<float> dg = {0.0, 0.0}; // re,im
+    default_gains = config.get_default<std::vector<float>>(unique_name, "frb_missing_gains", dg);
+
+    metadata_buf = get_buffer("in_buf");
+    register_consumer(metadata_buf, unique_name.c_str());
+    metadata_buffer_id = 0;
+    metadata_buffer_precondition_id = 0;
+    freq_idx = -1;
+    freq_MHz = -1;
+
+    // Gain for FRB
+    gain_frb_buf = get_buffer("gain_frb_buf");
+    gain_frb_buf_id = 0;
+    register_producer(gain_frb_buf, unique_name.c_str());
+    update_gains_frb = true;
+
+    // Gain for PSR
+    gain_psr_buf = get_buffer("gain_psr_buf");
+    gain_psr_buf_id = 0;
+    register_producer(gain_psr_buf, unique_name.c_str());
+    update_gains_psr = true;
+
+
+    using namespace std::placeholders;
+
+    // listen for gain updates FRB
+    string gainfrb = config.get_default<std::string>(unique_name, "updatable_config/gain_frb", "");
+    if (gainfrb.length() > 0)
+        configUpdater::instance().subscribe(
+            gainfrb, std::bind(&ReadGain::update_gains_frb_callback, this, _1));
+
+    // listen for gain updates PSR
+    string gainpsr = config.get<std::string>(unique_name, "updatable_config/gain_psr");
+    if (gainpsr.length() > 0)
+        configUpdater::instance().subscribe(
+            gainpsr, std::bind(&ReadGain::update_gains_psr_callback, this, _1));
+}
+
+bool ReadGain::update_gains_frb_callback(nlohmann::json& json) {
+    if (update_gains_frb) {
+        WARN("[FRB] cannot handle two back-to-back gain updates, rejecting the latter");
+        return false;
```

Comment:
Note that returning `false` here will exit kotekan.  I think given that we've listed this an a `WARN` condition.  We either need to change it for `ERROR` if we want this to kill the system, or return true, and just not apply the gains but keep running.   My impression is we don't want to cause instability, so in this case we should do the slightly bad thing and just return `true`, so we don't impact stability.   However we should be careful we don't get into a situation where it is possible to hit the condition and miss an update.